sympy/tensor/array/array_comprehension.py

```diff
@@ -253,27 +253,26 @@ def doit(self):
         if not self.is_shape_numeric:
             return self
 
-        return ImmutableDenseNDimArray(self._expand_array())
+        return self._expand_array()
 
     def _expand_array(self):
-        # To perform a subs at every element of the array.
-        def _array_subs(arr, var, val):
-            arr = MutableDenseNDimArray(arr)
-            for i in range(len(arr)):
-                index = arr._get_tuple_index(i)
-                arr[index] = arr[index].subs(var, val)
-            return arr.tolist()
-
-        list_gen = self.function
-        for var, inf, sup in reversed(self._limits):
-            list_expr = list_gen
-            list_gen = []
-            for val in range(inf, sup+1):
-                if not isinstance(list_expr, Iterable):
-                    list_gen.append(list_expr.subs(var, val))
-                else:
-                    list_gen.append(_array_subs(list_expr, var, val))
-        return list_gen
+        res = []
+        for values in itertools.product(*[range(inf, sup+1)
+                                        for var, inf, sup
+                                        in self._limits]):
+            if hasattr(self, '_lambda'):
```

Comment:
Let's make it like this: call a method (e.g. `res.append(self._get_element(values))`). Then you can overwrite this method in the subclass.